shared/src/main/scala/async/Semaphore.scala

```diff
@@ -0,0 +1,67 @@
+package gears.async
+
+import java.util.concurrent.ConcurrentLinkedQueue
+import java.util.concurrent.atomic.AtomicInteger
+
+/** A semaphore that manages a number of grants. One can wait to obtain a grant (with [[acquire]]) and return it to the
+  * semaphore (with [[release]]).
+  *
+  * @param value
+  *   the grant counter
+  */
+class Semaphore private (private var value: AtomicInteger) extends Async.Source[Unit]:
```

Comment:
If we want this to be a `Source` I think it's better to make it a source with a release value (a `Source[SemGuard]` with a `.release` method?) or maybe a `Resource[Unit]`?